### PR TITLE
Add 'len' mechanism for buffers 

### LINKED ISSUE #440: Need to create way to remove size on an input array from parameter list.
Right now we have a function such as the following:

`session.array_input_function(test_array_size, test_array)`

We really only want to do the following:

`session.array_input_function(test_array)`

Where in functions_addon.py we specify that test_array_size is len(test_array).

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
~~[ ] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~
### What does this Pull Request accomplish?
* Add 'len' mechanism for buffer length code generation
  * metadata looks like
    ```
        'ArrayInputFunction': { 'parameters': { 2: { 'size': {'mechanism':'len', 'value':'numberOfElements'}, }, }, },
    ```
  * This mechanism will set the size parameter to the length of the buffer parameter
* Created new filter function
  * Moved common filter code into shared function
* Changed 'skip_ivi_dance_size_parameter' to 'skip_size_parameter' in `ParameterUsageOptions`
  * More generic
### List issues fixed by this Pull Request below, if any.
* Fix #440 

### What testing has been done?
* Travis
* System tests

### CODE REVIEW

File: build/helper/codegen_helper.py [hunk 1]
```diff
@@ -113,14 +114,17 @@ def filter_parameters(function, parameter_usage_options):
     parameters_to_use = []
 
     # Filter based on options
-    ivi_dance_size_parameter = find_size_parameter(filter_ivi_dance_parameter(function['parameters']), function['parameters'])
+    # Find the size parameter - we are assuming there can only be one, eother from mechanism == 'ivi-dance' or mechanism == 'len'
```
Comment (marcoskirsch) [inline build/helper/codegen_helper.py]: typo

File: build/helper/metadata_filters.py [hunk 2]
```diff
@@ -21,15 +21,25 @@ def filter_enum_parameters(parameters):
     return [x for x in parameters if x['enum'] is not None]
 
 
-def filter_ivi_dance_parameter(parameters):
-    '''Returns the ivi-dance parameter of a session method if there is one. This is the parameter whose size is determined at runtime.'''
-    param = [x for x in parameters if x['size']['mechanism'] == 'ivi-dance']
-    assert len(param) <= 1, '{0} ivi-dance parameters. No more than one is allowed'.format(len(param))
+def filter_size_parameter(parameters, mechanism, direction=None):
```
Comment (marcoskirsch) [inline build/helper/metadata_filters.py]: please add docstring

File: build/helper/metadata_filters.py [hunk 3]
```diff
@@ -21,15 +21,25 @@ def filter_enum_parameters(parameters):
...
+    param = [x for x in parameters if x['size']['mechanism'] == mechanism]
+    assert len(param) <= 1, '{0} {1} parameters. No more than one is allowed'.format(len(param), mechanism)
     if len(param) == 0:
         return None
-    assert param[0]['direction'] == 'out', "ivi-dance parameter must have 'direction':'out'. Check your metadata."
-    assert param[0]['is_buffer'], "ivi-dance parameter must have 'is_buffer':True. Check your metadata."
+    if direction is not None:
+        assert param[0]['direction'] == direction, "{0} parameter must have 'direction':'out'. Check your metadata.".format(mechanism)
```
Comment (marcoskirsch) [inline build/helper/metadata_filters.py]: out is hard-coded in the format but it seems like it should be {1}

File: build/helper/metadata_filters.py (hunk 2 above)
Comment (marcoskirsch) [inline build/helper/metadata_filters.py]: prepend with _

File: build/helper/metadata_filters.py [hunk 4]
```diff
@@ -21,15 +21,25 @@ def filter_enum_parameters(parameters):
...
     return param[0]
 
 
+def filter_ivi_dance_parameter(parameters):
+    '''Returns the ivi-dance parameter of a session method if there is one. This is the parameter whose size is determined at runtime.'''
+    return filter_size_parameter(parameters, 'ivi-dance', 'out')
+
+
+def filter_len_parameter(parameters):
+    '''Returns the len parameter of a session method if there is one. This is the parameter whose size is determined at runtime.'''
```
Comment (marcoskirsch) [inline build/helper/metadata_filters.py]: I think len parameters are always direction "in". You should pass it to filter_size_parameter() and no longer have a default value for the parameter over there.

File: build/helper/codegen_helper.py (hunk 1 above)
Comment (texasaggie97) [reply build/helper/codegen_helper.py]: Fixed

File: build/helper/metadata_filters.py (hunk 2 above)
Comment (texasaggie97) [reply build/helper/metadata_filters.py]: Done

Comment (texasaggie97) [reply build/helper/metadata_filters.py]: Done

File: build/helper/metadata_filters.py (hunk 3 above)
Comment (texasaggie97) [reply build/helper/metadata_filters.py]: Fixed

File: build/helper/metadata_filters.py (hunk 4 above)
Comment (texasaggie97) [reply build/helper/metadata_filters.py]: I wasn't sure if they would always be in.

Changed

**Pull request merged**